### PR TITLE
reef: cephfs-journal-tool: Add preventive measures to avoid fs corruption

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/66278

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by joscollin** [comment]
@batrick 

This PR has a conflict because 728cdc349d0442926ab4f4401bd457d6ae75ca07 is not backported to reef.

Could you please review if [this](https://github.com/ceph/ceph/pull/57761/commits/b597ff504da7b32efebe6c9e1a8107f075efac56#diff-21e1004308449ef4639ea68b32b19951398157cece30ce6e7a48ab6b84c21456R141-R148) part of the code is good?

File: doc/cephfs/disaster-recovery-experts.rst [hunk 1]
```diff
@@ -68,9 +68,9 @@ truncate it like so:
 
 ::
 
-    cephfs-journal-tool [--rank=N] journal reset
+    cephfs-journal-tool [--rank=<fs_name>:{mds-rank|all}] journal reset --yes-i-really-really-mean-it
 
-Specify the MDS rank using the ``--rank`` option when the file system has/had
+Specify the filesystem and the MDS rank using the ``--rank`` option when the file system has/had
```
Comment (anthonyeleven) [inline doc/cephfs/disaster-recovery-experts.rst:73]: filesystem vs file system, let's be consistent.  I prefer the former but I think @zdover23 favors the latter.

File: src/tools/cephfs/JournalTool.cc [hunk 2]
```diff
@@ -251,14 +256,36 @@ int JournalTool::main_journal(std::vector<const char*> &argv)
     }
   } else if (command == "reset") {
     bool force = false;
-    if (argv.size() == 2) {
+    if (argv.size() == 1) {
+        std::cerr << "warning: this operation resets the journal!!!\n"
+                  << "Do not run this operation if you do not understand CephFS' internal storage mechanisms or have received specific instructions from those who do.\n"
```
Comment (anthonyeleven) [inline src/tools/cephfs/JournalTool.cc:261]: Only run this operation if you understand CephFS' internal storage mechanisms or have received specific instructions from those who do.

File: src/tools/cephfs/JournalTool.cc [hunk 3]
```diff
@@ -251,14 +256,36 @@ int JournalTool::main_journal(std::vector<const char*> &argv)
...
+                  << "If you want to continue, please add --yes-i-really-really-mean-it!!!"
+                  << std::endl;
+        return -EINVAL;
+    } else if (argv.size() == 2) {
+      if (std::string(argv[1]) == "--force") {
+        std::cerr << "warning: this operation resets the journal!!!\n"
+                  << "Do not run this operation if you do not understand CephFS' internal storage mechanisms or have received specific instructions from those who do.\n"
+                  << "If you want to continue, please add --yes-i-really-really-mean-it!!!"
```
Comment (anthonyeleven) [inline src/tools/cephfs/JournalTool.cc:269]: ditto

**Review by anthonyeleven — CHANGES_REQUESTED** [review]
There's wording that is counter to the intent, please adjust

File: src/tools/cephfs/JournalTool.cc (hunk 3 above)
Comment (joscollin) [reply src/tools/cephfs/JournalTool.cc:269]: I can't do code/doc changes in a backport. @anthonyeleven Please send a PR targeting to `main` branch or create a tracker issue for the changes you need.

Comment (anthonyeleven) [reply src/tools/cephfs/JournalTool.cc:269]: Hence my note to Zac.

**Comment by joscollin** [comment]
This backport has multiple conflicts as mentioned here: https://tracker.ceph.com/issues/66278. So changing the status to draft.

File: PendingReleaseNotes [hunk 4]
```diff
@@ -122,6 +122,11 @@
   cluster, MDS_CLIENTS_BROKEN_ROOTSQUASH. See the documentation on this warning
   and the new feature bit for more information.
 
+* cls_cxx_gather is marked as deprecated.
```
Comment (NitzanMordhai) [inline PendingReleaseNotes:125]: @joscollin ty for fixing that conflict, LGTM

**Review by NitzanMordhai — APPROVED** [review]
LGTM (the PendingReleaseNotes part)

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66766](https://tracker.ceph.com/issues/66766).

**Review by lxbsz — APPROVED** [review]
Finished checking all the failures, none is related to this PR. More detail please see the section "2024-07-16" in https://tracker.ceph.com/projects/cephfs/wiki/Reef.

**Pull request merged**